quantum/debounce/sym_defer_g.c

```diff
@@ -24,26 +24,39 @@ When no state changes have occured for DEBOUNCE milliseconds, we push the state.
 #    define DEBOUNCE 5
 #endif
 
+// Maximum debounce: 255ms
+#if DEBOUNCE > UINT8_MAX
+#    undef DEBOUNCE
+#    define DEBOUNCE UINT8_MAX
+#endif
+
+typedef uint8_t debounce_counter_t;
+
 #if DEBOUNCE > 0
-static bool         debouncing = false;
-static fast_timer_t debouncing_time;
+
+#    define DEBOUNCE_ELAPSED 0
+
+static debounce_counter_t debounce_counter = DEBOUNCE_ELAPSED;
+
+static fast_timer_t last_time;
 
 void debounce_init(uint8_t num_rows) {}
 
 bool debounce(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, bool changed) {
     bool cooked_changed = false;
 
     if (changed) {
-        debouncing      = true;
-        debouncing_time = timer_read_fast();
-    }
-
-    if (debouncing && timer_elapsed_fast(debouncing_time) >= DEBOUNCE) {
-        if (memcmp(cooked, raw, sizeof(matrix_row_t) * num_rows) != 0) {
-            memcpy(cooked, raw, sizeof(matrix_row_t) * num_rows);
-            cooked_changed = true;
+        debounce_counter = DEBOUNCE;
```

Comment:
The `debounce_counter` will only change between `DEBOUNCE_ELAPSED` and `DEBOUNCE`. Would a boolean variable (like the original) suffice?

```c
bool debounce(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, bool changed) {
    static fast_timer_t last_time;
    static bool debouncing = false;
    bool cooked_changed    = false;

    if (changed) {
        debouncing = true;
        last_time  = timer_read_fast();
    } else if (debouncing && timer_elapsed_fast(last_time) >= DEBOUNCE) {
        debouncing         = false;
        size_t matrix_size = num_rows * sizeof(matrix_row_t);
        if (memcmp(cooked, raw, matrix_size) != 0) {
            memcpy(cooked, raw, matrix_size);
            cooked_changed = true;
        }
    }

    return cooked_changed;
}
```

Reply:
Yes, but then the size of the binary goes back up. I'm not entirely sure of why and what sort of sorcery is behind the compiler's decision on that, but any operations involving booleans make the image a lot larger than using `uint_8` for the same purpose.

And since both data types are equally small and likely use the same pathways inside the controller, I would expect the performance to be the same.

Reply:
From my other comment:

> No idea why, but using a bool flag instead of the uint8_t counter makes it go from 16766 to 16814 bytes. Well, the increase if largely averted.

Reply:
In my own userspace (with presence of other and feature), the use of boolean resulted with a smaller firmware. They are likely nuances to optimisation applied with different combination of features and code.

That said, the larger output file difference seem to impact _only_ AVR with _direct pin_ configuration. They are a minority in the repository and I don't think it is worth obfuscating code readability for them.

Reply:
How much was the difference on the two boards you tested? And what are the command options you're using to compile each?

And I don't think it's obfuscating the readability so much, given it's exactly what the other implementations use as nomenclature and typing. Plus, if it ever becomes necessary to instead decrement the counter in steps, for whatever reason, this code is exactly 3 lines away from doing that. Not that it matters, but I like when every implementation follows the same pattern.

It'll be even more the case after my second PR, which will come after this, because it standardizes and cleans up all implementations.

Reply:
Just did a quick file size check using `a_dux`:
Current `sym_defer_g`:
```
Checking file size of a_dux_default.hex                                                             [OK]
 * The firmware size is fine - 19996/28672 (69%, 8676 bytes free)
```
Using `debounce_counter_t debounce_counter`:
```
Checking file size of a_dux_default.hex                                                             [OK]
 * The firmware size is fine - 20000/28672 (69%, 8672 bytes free)
```
Using `bool debouncing`:
```
Checking file size of a_dux_default.hex                                                             [OK]
 * The firmware size is fine - 19992/28672 (69%, 8680 bytes free)
```

This is also hard to explain.

Reply:
My guess is that it has something to do with the DEBOUNCE literal and the comparison with the function return value (alongside some weirdness around using the `bool` flags on the affected boards).

For whatever reason, it seems it's much more prone to optimizing stuff if it sees the same variable storing DEBOUNCE and using it to compare all the other stuff too.

Reply:
And the way I see it, there's a very, very minor increase in this case, but a "massive" one for the affected boards, so the lesser evil is likely to leave it as `uint8_t`.

Reply:
> but a "massive" one for the affected boards

I tested with `-kb cradio -km default` and `-kb a_dux -km default` on current `develop` with `avr-gcc 8.5.0`. Both are direct pin PCBs and the boolean variable resulted with more bytes free though.

Reply:
I'll retest tomorrow, but for cradio, those were the numbers I got.

Reply:
The difference on the boards that do have an increase with the `uint8_t` counter seem to be minimal (4~8 bytes) whereas for the boards where I saw an increase with the `bool` flag, it was by upwards of 100 bytes, so I think we should keep it this way.

Reply:
I setup a workflow using `qmkfm/qmk_cli` as a baseline to test different output, using `a_dux` as candidate:

- [develop](https://github.com/filterpaper/qmk_compiler/actions/runs/5958500297/job/16162818277#step:5:96) branch 8674 bytes free
- [sym_defer_g_bool](https://github.com/filterpaper/qmk_compiler/actions/runs/5958500297/job/16162818393#step:5:96) 8678 bytes free (src [diff](https://github.com/filterpaper/qmk_firmware/commit/209492f97f0351ae43000765e446ca9680900940))
- [sym_defer_g_uint8](https://github.com/filterpaper/qmk_compiler/actions/runs/5958500297/job/16162818508#step:5:96) 8670 bytes free (src [diff](https://github.com/filterpaper/qmk_firmware/commit/3c89e3d0a88adb688f2afe2a1b37be7fdf2721dd))

Reply:
Were all the numbers you sent earlier taken from compiling the code using each file as a custom debounce?

Reply:
> Were all the numbers you sent earlier taken from compiling the code using each file as a custom debounce?

Yes, they are sourced by the `rules.mk` file and the build action output confirms that the custom `debounce.c` is picked up by the compiler (see the earlier lines in the compile step output). I can see the same results locally but the QMK docker image has the official toolchain and is probably the best neutral resource to validate.

Reply:
Indeed, now the boolean flag is the smallest of them all, for all boards.

Something happened on `develop` in the meantime though. The firmware is a lot bigger now, in general. Before we were talking about 16.5KB and not it's around 20KB. 

I've reverted it to using a boolean flag now, to recover a few bytes.